### PR TITLE
Shell and Solid validation

### LINKED ISSUE #1594: Validate that coincident `HalfEdge`s refer to same `GlobalEdge`
If two `HalfEdge`s are coincident (which happens where two neighboring faces touch), they need to refer to the same `GlobalEdge`. If they don't, code working with the object graph (like the approximation code) won't know that they are supposed to be the same edge. In the case of the approximation code, this might lead to an invalid triangle mesh being generated. This has been the cause of bugs in the past (https://github.com/hannobraun/Fornjot/issues/494, https://github.com/hannobraun/Fornjot/issues/1162).

To make sure that all instances of this bug have been fixed, and that it stays fixed, we need a validation check for this.

### PR DESCRIPTION
**Pull request by A-Walrus**
Attempt to validate `Shell`'s by making sure that each half-edge in a shell refers to a global-edge that is referred to by another half-edge in that same shell. My logic says that each edge in a shell should be a part of exactly two faces but I may be wrong.

This validation check currently fails with `cargo run --package export- validator`, but I'm not sure if this is a bug in my validation checks, or whether I exposed a bug in existing code.

For example looking at the cuboid example, many global edges are referred to by only one half_edge, which seems wrong to me...

edit by @hannobraun:
Close #1594

### CODE REVIEW

**Comment by hannobraun** [comment]
Thank you, @A-Walrus! I'll take a closer look on Monday. For now, some comments.


> My logic says that each edge in a shell should be a part of exactly two faces but I may be wrong.

No, I think that should be correct! At least it matches what I've been thinking.


> For example looking at the cuboid example, many global edges are referred to by only one half_edge, which seems wrong to me...

Yes, that is definitely wrong. Unless there's a problem with the new validation code, this is likely a bug in the sweep code. I'm pretty sure that the sweep code creates the correct vertices (I've had problems with exactly that recently, and I'm pretty sure that's dealt with), but it might be creating duplicate `GlobalEdge`s. As long as those refer to the correct vertices, it's plausible that all the code we currently have still works correctly.

So yeah, totally plausible you exposed a bug!

**Comment by hannobraun** [comment]
I've taken a look at the sweep code, and I think I've found a bug:
https://github.com/hannobraun/Fornjot/blob/ccebda5e0adb248deb52aa51a1c63901540da27e/crates/fj-kernel/src/builder/cycle.rs#L48-L49

This is missing a call to `HalfEdgeBuilder::with_global_form`. That this call is missing definitely explains the validation failure.

Problem is though, if I add the call, validation still fails. I don't know what that means. Either there are more cases like this (I couldn't find any), or something weird is going on.

**Comment by A-Walrus** [comment]
There seems to have been a couple problems:
- What you found about `with_global_form`
- Incorrect order in `global_edges`
- Needing to cache edges like we do vertices

File: crates/fj-kernel/src/algorithms/sweep/mod.rs [hunk 1]
```diff
@@ -47,4 +47,6 @@ pub trait Sweep: Sized {
 pub struct SweepCache {
     /// Cache for global vertices
     pub global_vertex: BTreeMap<ObjectId, Handle<Vertex>>,
+    /// Cache for global edges
+    pub global_edge: BTreeMap<ObjectId, Handle<GlobalEdge>>,
```
Comment (A-Walrus) [inline crates/fj-kernel/src/algorithms/sweep/mod.rs:51]: Maybe we should have a more general caching solution. This seems to work for now but IDK if it's ideal.

Comment (hannobraun) [reply crates/fj-kernel/src/algorithms/sweep/mod.rs:51]: I don't know, this seems fine to me. If you come up with something better, feel free to suggest it of course, but I think we can keep it as-is.

File: crates/fj-kernel/src/algorithms/sweep/vertex.rs [hunk 2]
```diff
@@ -26,7 +26,11 @@ impl Sweep for Handle<Vertex> {
             .clone();
 
         let vertices = [a, b];
-        let global_edge = GlobalEdge::new().insert(objects);
+        let global_edge = cache
+            .global_edge
+            .entry(self.id())
+            .or_insert_with(|| GlobalEdge::new().insert(objects))
+            .clone();
```
Comment (hannobraun) [inline crates/fj-kernel/src/algorithms/sweep/vertex.rs:33]: I swear, I was looking for exactly this code yesterday. I found the vertex caching code instead, and my brain was like "oh, so we do have the required caching". I'm pretty stupid sometimes :smile: 

In any case, good find!

**Review by hannobraun — DISMISSED** [review]
Thank you, @A-Walrus, this is great!

> There seems to have been a couple problems:
> 
> * What you found about `with_global_form`
> * Incorrect order in `global_edges`
> * Needing to cache edges like we do vertices

This is some nice debugging. Well done!

I left some comments, but there's nothing critical. There's one suggestion that I will apply myself after submitting this review. This is good to merge!

Some additional comments:

- In general, I'd prefer if the final order of commits were such, that every single commit works (that means moving the fixes before the validations that require them). This doesn't matter most of the time, but if you're using `git bisect`, then a commit that has build errors or test failures can be inconvenient.
- As you mention, the test cases only cover failure. This would be hard to fix right now, as we don't have good tools to build `Shell`s and `Solid`s. But since I'm going to [focus on APIs to build shapes](https://github.com/hannobraun/Fornjot/issues/42#issuecomment-1473709945) for the foreseeable future, those test cases are probably a good testing ground for those new APIs.

**Pull request merged**